### PR TITLE
Locking

### LINKED ISSUE #107: Improve selection by ignoring locked game objects
Objects previously were not selectable if they had a:
- Static flag
- Marked as locked

However, the locked status only lasted for a single editing session and did not persist. Now, locking actually locks the object and persists that in the scene, which means that it is not selectable by ray or direct selection and will show as locked in the Hierarchy View in the 2D editor.

### LINKED ISSUE #108: Locked Objects Workspace
With the improvement to locking (#107) there is now a new Locked Objects Workspace to make it easier to see what objects are locked in the scene. You can still use the Hierarchy Workspace and filter for Locked objects as well.

### PR DESCRIPTION
**Pull request by amirebrahimi**
Resolves #107 
Closes #108 

@timoni - I've added you to this in case you'd like to try it out. Please approve if you do or put comments for anything that needs addressing.

### CODE REVIEW

File: Workspaces/Hierarchy/Scripts/HierarchyListViewController.cs [hunk 1]
```diff
@@ -154,13 +169,27 @@ protected override void UpdateRecursively(List<HierarchyData> data, ref float of
...
-				if (hasFilterQuery)
+
+				if (hasLockedQuery || hasFilterQuery)
 				{
-					var filterTestPass = datum.types.Any(type => matchesFilter(type));
+					var filterTestPass = true;
 
-					if (!filterTestPass) // If this item doesn't match the filter, move on to the next item; do not count
+					if (hasLockedQuery)
+						filterTestPass = isLocked((GameObject)EditorUtility.InstanceIDToObject(datum.index));
```
Comment (mtschoen-unity) [inline Workspaces/Hierarchy/Scripts/HierarchyListViewController.cs]: This kind of thing could be used in UpdateSelf to toggle the lock icon

File: Workspaces/Inspector/Scripts/InspectorListViewController.cs [hunk 2]
```diff
@@ -180,6 +180,15 @@ void UpdateItem(Transform t, float offset, bool dontSettle, ref bool doneSettlin
 			UpdateItemTransform(t, targetPosition, targetRotation, dontSettle, ref doneSettling);
 		}
 
+		protected override void RecycleItem(string template, InspectorListItem item)
+		{
+			var headerItem = item as InspectorHeaderItem;
+			if (headerItem)
+				headerItem.setLocked = null; // Reset, so it doesn't get called for the wrong object when re-used
```
Comment (mtschoen-unity) [inline Workspaces/Inspector/Scripts/InspectorListViewController.cs]: Are you sure this is necessary? In what case would a HeaderItem not get this overwritten in GetItem?

File: Workspaces/Inspector/Scripts/InspectorListViewController.cs [hunk 3]
```diff
@@ -205,8 +214,8 @@ protected override InspectorListItem GetItem(InspectorData listData)
 			if (headerItem)
 			{
 				var go = (GameObject)listData.serializedObject.targetObject;
-				headerItem.lockToggle.isOn = isLocked(go);
 				headerItem.setLocked = locked => setLocked(go, locked);
+				headerItem.lockToggle.isOn = isLocked(go);
```
Comment (mtschoen-unity) [inline Workspaces/Inspector/Scripts/InspectorListViewController.cs]: Why this change?

File: Workspaces/LockedObjects/LockedObjectsWorkspace.cs [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+using System;
+using System.Collections.Generic;
+using UnityEditor.Experimental.EditorVR.Utilities;
+using UnityEngine;
+using UnityEngine.UI;
+
+namespace UnityEditor.Experimental.EditorVR.Workspaces
+{
+	[MainMenuItem("Locked Objects", "Workspaces", "View all locked objects in your scene(s)")]
+	class LockedObjectsWorkspace : HierarchyWorkspace, IUsesGameObjectLocking
```
Comment (mtschoen-unity) [inline Workspaces/LockedObjects/LockedObjectsWorkspace.cs]: Nice. I was hoping this would work 👍

File: Scripts/Data/HierarchyData.cs [hunk 5]
```diff
@@ -17,13 +17,15 @@ public override int index
 		public int instanceID { private get; set; }
 
 		public HashSet<string> types { get; set; }
+		public bool locked { get; set; }
```
Comment (dunity) [inline Scripts/Data/HierarchyData.cs]: Agreed with @mtschoen on this point.  I'd vote to keep the state outside of the model in this case too.

File: Workspaces/LockedObjects/LockedObjectsWorkspace.cs [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+	{
+		[SerializeField]
+		GameObject m_UnlockAllPrefab;
+
+		string baseSearchQuery;
```
Comment (dunity) [inline Workspaces/LockedObjects/LockedObjectsWorkspace.cs]: m_BaseSearchQuery

File: Workspaces/LockedObjects/LockedObjectsWorkspace.cs [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+		{
+			base.Setup();
+
+			if (m_UnlockAllPrefab)
+			{
+				var unlockAllUI = ObjectUtils.Instantiate(m_UnlockAllPrefab, m_WorkspaceUI.frontPanel, false);
+				foreach (var mb in unlockAllUI.GetComponentsInChildren<MonoBehaviour>())
+				{
+					connectInterfaces(mb);
+				}
```
Comment (dunity) [inline Workspaces/LockedObjects/LockedObjectsWorkspace.cs]: Add new line below

File: Workspaces/Inspector/Scripts/InspectorListViewController.cs (hunk 2 above)
Comment (dunity) [inline Workspaces/Inspector/Scripts/InspectorListViewController.cs]: Only InspectorHeaderItems get setLocked toggled in GetItem.  Seems like this would still be needed here specifically for InspectorListItems.

File: Workspaces/Inspector/Scripts/InspectorListViewController.cs (hunk 3 above)
Comment (amirebrahimi) [inline Workspaces/Inspector/Scripts/InspectorListViewController.cs]: Fix for bug with switching objects in the HierarchyWorkspace would unset a lock on a previous game object (from the closure). Setting the callback first fixes this.

**Comment by amirebrahimi** [comment]
@mtschoen-unity / @dunity - Addressed your comments.
